src/Authentication.Abstractions/AzureEnvironment.cs

```diff
@@ -131,18 +153,185 @@ public class AzureEnvironment : IAzureEnvironment
             result[EnvironmentName.AzureCloud] = azureCloud;
             result[EnvironmentName.AzureChinaCloud] = azureChina;
             result[EnvironmentName.AzureUSGovernment] = azureUSGovernment;
             result[EnvironmentName.AzureGermanCloud] = azureGermany;
+            SetExtendedProperties(result);
+            return result;
+        }
+
+        /// <summary>
+        /// Initializes cloud metadata dynamically from ARM.
+        /// </summary>
+        private static async Task<IDictionary<string, AzureEnvironment>> InitializeEnvironmentsFromArm(string armMetadataRequestUri)
+        {
+            HttpResponseMessage armResponseMessage = null;
+            await Policy
+                .Handle<HttpRequestException>()
+                .WaitAndRetryAsync(
+                    GetArmMetadataRetryCount,
+                    r => TimeSpan.FromSeconds(2 * r),
+                    (ex, ts, r) =>
+                        Debug.WriteLine(
+                            $"Could not retrieve ARM metadata. Attempt #{r}/{GetArmMetadataRetryCount}. Will try again in {ts.TotalSeconds} seconds. Exception={ex}"))
+                .ExecuteAsync(async () =>
+                {
+                    armResponseMessage = await Client.GetAsync(armMetadataRequestUri);
+                });
+
+            if (armResponseMessage == null || armResponseMessage?.StatusCode != HttpStatusCode.OK)
+            {
+                throw new Exception("Could not retrieve ARM metadata, compute management client cannot be initialized.");
+            }
+
+            var armMetadataContent = await armResponseMessage.Content?.ReadAsStringAsync();
+            if (string.IsNullOrEmpty(armMetadataContent))
+            {
+                throw new Exception("Could not read ARM metadata, compute management client cannot be initialized.");
+            }
+
+            var armMetadataList = JsonConvert.DeserializeObject<List<ArmMetadata>>(armMetadataContent);
+            var result = new ConcurrentDictionary<string, AzureEnvironment>(StringComparer.InvariantCultureIgnoreCase);
+            foreach (var armMetadata in armMetadataList)
+            {
+                result[armMetadata.Name] = MapArmToAzureEnvironment(armMetadataList.First(a =>
+                    a.Name.Equals(armMetadata.Name, StringComparison.InvariantCultureIgnoreCase)));
+            }
 
+            SetExtendedProperties(result);
             return result;
         }
 
+        /// <summary>
+        /// Set Extended properties (to maintain parity).
+        /// </summary>
+        /// <param name="azureEnvironments">Collection of AzureEnvironments</param>
+        private static void SetExtendedProperties(IDictionary<string, AzureEnvironment> azureEnvironments)
+        {
+            azureEnvironments[EnvironmentName.AzureCloud].SetProperty(ExtendedEndpoint.OperationalInsightsEndpoint, AzureEnvironmentConstants.AzureOperationalInsightsEndpoint);
```

Comment:
Is it guaranteed ` azureEnvironments[EnvironmentName.***Cloud]` is not null?

Reply:
Yes. EnvironmentName is a static class in Microsoft.Azure.Commands.Common.Authentication.Abstractions, and the cloud names are constant strings.